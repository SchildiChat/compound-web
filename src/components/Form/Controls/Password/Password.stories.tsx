/*
Copyright 2023 The Matrix.org Foundation C.I.C.
Copyright 2023 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from "react";

import { Meta, StoryObj } from "@storybook/react";

import { PasswordInput } from "./";
import { within } from "@storybook/test";
import { userEvent } from "@storybook/test";

type Props = { invalid?: boolean } & React.ComponentProps<typeof PasswordInput>;

export default {
  title: "Form/Controls/Password",
  component: PasswordInput,
  tags: ["autodocs"],
  parameters: {
    controls: {
      include: [
        "defaultValue",
        "placeholder",
        "disabled",
        "autoFocus",
        "readOnly",
        "invalid",
      ],
    },
  },
  argTypes: {
    defaultValue: {
      type: "string",
    },
    placeholder: {
      type: "string",
    },
    disabled: {
      type: "boolean",
    },
    autoFocus: {
      type: "boolean",
    },
    readOnly: {
      type: "boolean",
    },
    invalid: {
      type: "boolean",
    },
  },
  render: ({ invalid, ...restArgs }) => (
    <PasswordInput data-invalid={invalid || undefined} {...restArgs} />
  ),
  args: {
    placeholder: "",
    autoFocus: false,
    disabled: false,
    readOnly: false,
    invalid: false,
  },
} satisfies Meta<Props>;

type Story = StoryObj<Props>;

export const Empty: Story = {
  parameters: {
    design: {
      type: "figma",
      url: "https://www.figma.com/file/rTaQE2nIUSLav4Tg3nozq7/Compound-Web-Components?type=design&node-id=793-4477",
    },
  },
};

export const Filled: Story = {
  args: {
    defaultValue: "Filled",
  },
  parameters: {
    design: {
      type: "figma",
      url: "https://www.figma.com/file/rTaQE2nIUSLav4Tg3nozq7/Compound-Web-Components?type=design&node-id=793-4536",
    },
  },
};

export const Disabled: Story = {
  args: {
    defaultValue: "Disabled",
    disabled: true,
  },
  parameters: {
    design: {
      type: "figma",
      url: "https://www.figma.com/file/rTaQE2nIUSLav4Tg3nozq7/Compound-Web-Components?type=design&node-id=793-4530",
    },
  },
};

export const ReadOnly: Story = {
  args: {
    defaultValue: "Read only",
    readOnly: true,
  },
  parameters: {
    design: {
      type: "figma",
      url: "https://www.figma.com/file/rTaQE2nIUSLav4Tg3nozq7/Compound-Web-Components?type=design&node-id=2126-21984",
    },
  },
};

export const Focused: Story = {
  args: {
    autoFocus: true,
  },
};

export const Invalid: Story = {
  args: {
    defaultValue: "Invalid",
    invalid: true,
  },
  parameters: {
    design: {
      type: "figma",
      url: "https://www.figma.com/file/rTaQE2nIUSLav4Tg3nozq7/Compound-Web-Components?type=design&node-id=793-4514",
    },
  },
};

export const Visible: Story = {
  args: {
    defaultValue: "hunter2",
  },
  parameters: {
    design: {
      type: "figma",
      url: "https://www.figma.com/file/rTaQE2nIUSLav4Tg3nozq7/Compound-Web-Components?type=design&node-id=793-4536",
    },
  },
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement);
    const toggle = canvas.getByLabelText("Show");
    await userEvent.click(toggle);
  },
};
