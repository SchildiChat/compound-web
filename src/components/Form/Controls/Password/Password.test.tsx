/*
Copyright 2023 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { describe, it, expect } from "vitest";
import React from "react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { PasswordInput } from "./Password";

describe("PasswordControl", () => {
  it("switches the input type", async () => {
    const user = userEvent.setup();
    const { container } = render(<PasswordInput defaultValue="p4ssw0rd!" />);

    expect(container.querySelector("[type=password]")).toBeInTheDocument();
    expect(container).toMatchSnapshot("invisible");

    await user.click(screen.getByRole("button", { name: "Show" }));

    expect(container.querySelector("[type=password]")).not.toBeInTheDocument();
    expect(container.querySelector("[type=text]")).toBeInTheDocument();
    expect(container).toMatchSnapshot("visible");

    await user.click(screen.getByRole("button", { name: "Hide" }));

    expect(container.querySelector("[type=password]")).toBeInTheDocument();
    expect(container).toMatchSnapshot("invisible");
  });
});
