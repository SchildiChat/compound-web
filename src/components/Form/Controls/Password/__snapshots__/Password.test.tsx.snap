// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PasswordControl > switches the input type > invisible 1`] = `
<div>
  <div
    class="_container_1689d3"
    id=":r0:"
  >
    <input
      class="_control_a839aa _control_1689d3"
      type="password"
      value="p4ssw0rd!"
    />
    <button
      aria-controls=":r0:"
      aria-label="Show"
      class="_action_1689d3"
      data-state="closed"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="cpd-icon"
        fill="none"
        height="1em"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M16.1 13.3L14.65 11.85C14.8 11.0666 14.575 10.3333 13.975 9.64998C13.375 8.96664 12.6 8.69998 11.65 8.84998L10.2 7.39998C10.4834 7.26664 10.7709 7.16664 11.0625 7.09998C11.3542 7.03331 11.6667 6.99998 12 6.99998C13.25 6.99998 14.3125 7.43748 15.1875 8.31248C16.0625 9.18748 16.5 10.25 16.5 11.5C16.5 11.8333 16.4667 12.1458 16.4 12.4375C16.3334 12.7291 16.2334 13.0166 16.1 13.3ZM19.3 16.45L17.85 15.05C18.4834 14.5666 19.0459 14.0375 19.5375 13.4625C20.0292 12.8875 20.45 12.2333 20.8 11.5C19.9667 9.81664 18.7709 8.47914 17.2125 7.48748C15.6542 6.49581 13.9167 5.99998 12 5.99998C11.5167 5.99998 11.0417 6.03331 10.575 6.09998C10.1084 6.16664 9.65005 6.26664 9.20005 6.39998L7.65005 4.84998C8.33338 4.56664 9.03338 4.35414 9.75005 4.21248C10.4667 4.07081 11.2167 3.99998 12 3.99998C14.3834 3.99998 16.525 4.62914 18.425 5.88748C20.325 7.14581 21.75 8.78331 22.7 10.8C22.75 10.8833 22.7834 10.9875 22.8 11.1125C22.8167 11.2375 22.825 11.3666 22.825 11.5C22.825 11.6333 22.8125 11.7625 22.7875 11.8875C22.7625 12.0125 22.7334 12.1166 22.7 12.2C22.3167 13.05 21.8375 13.8333 21.2625 14.55C20.6875 15.2666 20.0334 15.9 19.3 16.45ZM19.1 21.9L15.6 18.45C15.0167 18.6333 14.4292 18.7708 13.8375 18.8625C13.2459 18.9541 12.6334 19 12 19C9.61672 19 7.47505 18.3708 5.57505 17.1125C3.67505 15.8541 2.25005 14.2166 1.30005 12.2C1.25005 12.1166 1.21672 12.0125 1.20005 11.8875C1.18338 11.7625 1.17505 11.6333 1.17505 11.5C1.17505 11.3666 1.18338 11.2416 1.20005 11.125C1.21672 11.0083 1.25005 10.9083 1.30005 10.825C1.65005 10.075 2.06672 9.38331 2.55005 8.74998C3.03338 8.11664 3.56672 7.53331 4.15005 6.99998L2.07505 4.89998C1.89172 4.71664 1.80005 4.48748 1.80005 4.21248C1.80005 3.93748 1.90005 3.69998 2.10005 3.49998C2.28338 3.31664 2.51672 3.22498 2.80005 3.22498C3.08338 3.22498 3.31672 3.31664 3.50005 3.49998L20.5 20.5C20.6834 20.6833 20.7792 20.9125 20.7875 21.1875C20.7959 21.4625 20.7 21.7 20.5 21.9C20.3167 22.0833 20.0834 22.175 19.8 22.175C19.5167 22.175 19.2834 22.0833 19.1 21.9ZM5.55005 8.39998C5.06672 8.83331 4.62505 9.30831 4.22505 9.82498C3.82505 10.3416 3.48338 10.9 3.20005 11.5C4.03338 13.1833 5.22922 14.5208 6.78755 15.5125C8.34588 16.5041 10.0834 17 12 17C12.3334 17 12.6584 16.9791 12.975 16.9375C13.2917 16.8958 13.6167 16.85 13.95 16.8L13.05 15.85C12.8667 15.9 12.6917 15.9375 12.525 15.9625C12.3584 15.9875 12.1834 16 12 16C10.75 16 9.68755 15.5625 8.81255 14.6875C7.93755 13.8125 7.50005 12.75 7.50005 11.5C7.50005 11.3166 7.51255 11.1416 7.53755 10.975C7.56255 10.8083 7.60005 10.6333 7.65005 10.45L5.55005 8.39998Z"
          fill="currentColor"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`PasswordControl > switches the input type > invisible 2`] = `
<div>
  <div
    class="_container_1689d3"
    id=":r0:"
  >
    <input
      class="_control_a839aa _control_1689d3"
      type="password"
      value="p4ssw0rd!"
    />
    <button
      aria-controls=":r0:"
      aria-label="Show"
      class="_action_1689d3"
      data-state="closed"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="cpd-icon"
        fill="none"
        height="1em"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M16.1 13.3L14.65 11.85C14.8 11.0666 14.575 10.3333 13.975 9.64998C13.375 8.96664 12.6 8.69998 11.65 8.84998L10.2 7.39998C10.4834 7.26664 10.7709 7.16664 11.0625 7.09998C11.3542 7.03331 11.6667 6.99998 12 6.99998C13.25 6.99998 14.3125 7.43748 15.1875 8.31248C16.0625 9.18748 16.5 10.25 16.5 11.5C16.5 11.8333 16.4667 12.1458 16.4 12.4375C16.3334 12.7291 16.2334 13.0166 16.1 13.3ZM19.3 16.45L17.85 15.05C18.4834 14.5666 19.0459 14.0375 19.5375 13.4625C20.0292 12.8875 20.45 12.2333 20.8 11.5C19.9667 9.81664 18.7709 8.47914 17.2125 7.48748C15.6542 6.49581 13.9167 5.99998 12 5.99998C11.5167 5.99998 11.0417 6.03331 10.575 6.09998C10.1084 6.16664 9.65005 6.26664 9.20005 6.39998L7.65005 4.84998C8.33338 4.56664 9.03338 4.35414 9.75005 4.21248C10.4667 4.07081 11.2167 3.99998 12 3.99998C14.3834 3.99998 16.525 4.62914 18.425 5.88748C20.325 7.14581 21.75 8.78331 22.7 10.8C22.75 10.8833 22.7834 10.9875 22.8 11.1125C22.8167 11.2375 22.825 11.3666 22.825 11.5C22.825 11.6333 22.8125 11.7625 22.7875 11.8875C22.7625 12.0125 22.7334 12.1166 22.7 12.2C22.3167 13.05 21.8375 13.8333 21.2625 14.55C20.6875 15.2666 20.0334 15.9 19.3 16.45ZM19.1 21.9L15.6 18.45C15.0167 18.6333 14.4292 18.7708 13.8375 18.8625C13.2459 18.9541 12.6334 19 12 19C9.61672 19 7.47505 18.3708 5.57505 17.1125C3.67505 15.8541 2.25005 14.2166 1.30005 12.2C1.25005 12.1166 1.21672 12.0125 1.20005 11.8875C1.18338 11.7625 1.17505 11.6333 1.17505 11.5C1.17505 11.3666 1.18338 11.2416 1.20005 11.125C1.21672 11.0083 1.25005 10.9083 1.30005 10.825C1.65005 10.075 2.06672 9.38331 2.55005 8.74998C3.03338 8.11664 3.56672 7.53331 4.15005 6.99998L2.07505 4.89998C1.89172 4.71664 1.80005 4.48748 1.80005 4.21248C1.80005 3.93748 1.90005 3.69998 2.10005 3.49998C2.28338 3.31664 2.51672 3.22498 2.80005 3.22498C3.08338 3.22498 3.31672 3.31664 3.50005 3.49998L20.5 20.5C20.6834 20.6833 20.7792 20.9125 20.7875 21.1875C20.7959 21.4625 20.7 21.7 20.5 21.9C20.3167 22.0833 20.0834 22.175 19.8 22.175C19.5167 22.175 19.2834 22.0833 19.1 21.9ZM5.55005 8.39998C5.06672 8.83331 4.62505 9.30831 4.22505 9.82498C3.82505 10.3416 3.48338 10.9 3.20005 11.5C4.03338 13.1833 5.22922 14.5208 6.78755 15.5125C8.34588 16.5041 10.0834 17 12 17C12.3334 17 12.6584 16.9791 12.975 16.9375C13.2917 16.8958 13.6167 16.85 13.95 16.8L13.05 15.85C12.8667 15.9 12.6917 15.9375 12.525 15.9625C12.3584 15.9875 12.1834 16 12 16C10.75 16 9.68755 15.5625 8.81255 14.6875C7.93755 13.8125 7.50005 12.75 7.50005 11.5C7.50005 11.3166 7.51255 11.1416 7.53755 10.975C7.56255 10.8083 7.60005 10.6333 7.65005 10.45L5.55005 8.39998Z"
          fill="currentColor"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`PasswordControl > switches the input type > visible 1`] = `
<div>
  <div
    class="_container_1689d3"
    id=":r0:"
  >
    <input
      class="_control_a839aa _control_1689d3"
      type="text"
      value="p4ssw0rd!"
    />
    <button
      aria-controls=":r0:"
      aria-label="Hide"
      class="_action_1689d3"
      data-state="closed"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="cpd-icon"
        fill="none"
        height="1em"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12 16C13.25 16 14.3125 15.5625 15.1875 14.6875C16.0625 13.8125 16.5 12.75 16.5 11.5C16.5 10.25 16.0625 9.1875 15.1875 8.3125C14.3125 7.4375 13.25 7 12 7C10.75 7 9.68755 7.4375 8.81255 8.3125C7.93755 9.1875 7.50005 10.25 7.50005 11.5C7.50005 12.75 7.93755 13.8125 8.81255 14.6875C9.68755 15.5625 10.75 16 12 16ZM12 14.2C11.25 14.2 10.6125 13.9375 10.0875 13.4125C9.56255 12.8875 9.30005 12.25 9.30005 11.5C9.30005 10.75 9.56255 10.1125 10.0875 9.5875C10.6125 9.0625 11.25 8.8 12 8.8C12.75 8.8 13.3875 9.0625 13.9125 9.5875C14.4375 10.1125 14.7 10.75 14.7 11.5C14.7 12.25 14.4375 12.8875 13.9125 13.4125C13.3875 13.9375 12.75 14.2 12 14.2ZM12 19C9.68338 19 7.56672 18.3875 5.65005 17.1625C3.73338 15.9375 2.28338 14.2833 1.30005 12.2C1.25005 12.1167 1.21672 12.0125 1.20005 11.8875C1.18338 11.7625 1.17505 11.6333 1.17505 11.5C1.17505 11.3667 1.18338 11.2375 1.20005 11.1125C1.21672 10.9875 1.25005 10.8833 1.30005 10.8C2.28338 8.71667 3.73338 7.0625 5.65005 5.8375C7.56672 4.6125 9.68338 4 12 4C14.3167 4 16.4334 4.6125 18.35 5.8375C20.2667 7.0625 21.7167 8.71667 22.7 10.8C22.75 10.8833 22.7834 10.9875 22.8 11.1125C22.8167 11.2375 22.825 11.3667 22.825 11.5C22.825 11.6333 22.8167 11.7625 22.8 11.8875C22.7834 12.0125 22.75 12.1167 22.7 12.2C21.7167 14.2833 20.2667 15.9375 18.35 17.1625C16.4334 18.3875 14.3167 19 12 19ZM12 17C13.8834 17 15.6125 16.5042 17.1875 15.5125C18.7625 14.5208 19.9667 13.1833 20.8 11.5C19.9667 9.81667 18.7625 8.47917 17.1875 7.4875C15.6125 6.49583 13.8834 6 12 6C10.1167 6 8.38755 6.49583 6.81255 7.4875C5.23755 8.47917 4.03338 9.81667 3.20005 11.5C4.03338 13.1833 5.23755 14.5208 6.81255 15.5125C8.38755 16.5042 10.1167 17 12 17Z"
          fill="currentColor"
        />
      </svg>
    </button>
  </div>
</div>
`;
